crates/sats/src/slim_slice.rs

```diff
@@ -0,0 +1,1060 @@
+    slice,
+    str::{from_utf8_unchecked, from_utf8_unchecked_mut},
+};
+use thiserror::Error;
+
+// =============================================================================
+// Errors
+// =============================================================================
+
+#[derive(Error, Debug)]
+#[error("The length `{len}` was too long")]
+/// An error signifying that a container's size was over `u32::MAX`.
+pub struct LenTooLong<T = ()> {
+    /// The size of the container that was too large.
+    pub len: usize,
+    /// The container that was too large for into-slim conversion.
+    pub too_long: T,
+}
+
+impl<T> LenTooLong<T> {
+    /// Forgets the container part of the error.
+    pub fn forget(self) -> LenTooLong {
+        self.map(drop)
+    }
+
+    /// Maps the container part of the error.
+    pub fn map<U>(self, with: impl FnOnce(T) -> U) -> LenTooLong<U> {
+        LenTooLong {
+            len: self.len,
+            too_long: with(self.too_long),
+        }
+    }
+}
+
+/// Try to convert `x` into `B` and forget the container part of the error if any.
+pub fn try_into<A, B: TryFrom<A, Error = LenTooLong<A>>>(x: A) -> Result<B, LenTooLong> {
+    x.try_into().map_err(|e: LenTooLong<A>| e.forget())
+}
+
+impl<A> crate::ser::Error for LenTooLong<A> {
+    fn custom<T: Display>(_: T) -> Self {
+        unimplemented!()
+    }
+}
+
+// =============================================================================
+// Utils
+// =============================================================================
+
+/// Ensures that `$thing.len() <= u32::MAX`.
+macro_rules! ensure_len_fits {
+    ($thing:expr) => {
+        let Ok(_) = u32::try_from($thing.len()) else {
+            return Err(LenTooLong {
+                len: $thing.len(),
+                too_long: $thing,
+            });
+        };
+    };
+}
+
+/// Convert to `A` but panic if `x.len() > u32::MAX`.
+fn expect_fit<A, E, B: TryInto<A, Error = LenTooLong<E>>>(x: B) -> A {
+    x.try_into().map_err(|e| e.len).expect("length didn't fit in `u32`")
+}
+
+fn into_box<T, U: Into<Box<[T]>>>(x: U) -> Box<[T]> {
+    x.into()
+}
+
+// =============================================================================
+// Raw slice utility type
+// =============================================================================
+
+/// Implementation detail of the other types.
+/// Provides some convenience but users of the type are responsible
+/// for safety, invariants and variance.
+#[repr(packed)]
+struct SlimRawSlice<T> {
+    /// The length of the slice.
+    len: u32,
+    /// A valid pointer to the slice data.
+    ptr: NonNull<T>,
+}
+
+impl<T> SlimRawSlice<T> {
+    /// Split `self` into a raw pointer and the slice length.
+    fn split(self) -> (*mut T, usize) {
+        (self.ptr.as_ptr(), self.len as usize)
+    }
+
+    /// Dereferences this raw slice into a shared slice.
+    ///
+    /// SAFETY: `self.ptr` and `self.len` must satisfy `from_raw_parts`'s requirements.
```

Comment:
Could you maybe explicitly say that it is slice's requirements? ie that `from_raw_parts` is not referring to a method on SlimRawSlice?

Reply:
In particular I think spelling out those requirements would be good, even if they are listed out in core::slice docs or wherever.